src/endpoint.jl

```diff
@@ -76,96 +122,78 @@ Visualizing two contiguous intervals can assist in understanding this logic:
+        isunbounded(a) && isunbounded(b) && isleft(a) == isleft(b)||
+        a.endpoint == b.endpoint && 
+        (isleft(a) == isleft(b) ? isclosed(a) == isclosed(b) :
+             isclosed(a) && isclosed(b))
     )
 end
 
-function Base.isequal(a::LeftEndpoint, b::RightEndpoint)
-    isequal(a.endpoint, b.endpoint) && isclosed(a) && isclosed(b)
-end
-
-function Base.isequal(a::RightEndpoint, b::LeftEndpoint)
-    isequal(b, a)
-end
-
-function Base.isless(a::LeftEndpoint, b::LeftEndpoint)
-    return (
-        !isunbounded(b) && (
-            isunbounded(a) ||
-            a.endpoint < b.endpoint ||
-            a.endpoint == b.endpoint && isclosed(a) && !isclosed(b)
-        )
-    )
-end
-
-function Base.isless(a::RightEndpoint, b::RightEndpoint)
-    return (
-        !isunbounded(a) && (
-            isunbounded(b) ||
-            a.endpoint < b.endpoint ||
-            a.endpoint == b.endpoint && !isclosed(a) && isclosed(b)
-        )
-    )
-end
-
-function Base.isless(a::LeftEndpoint, b::RightEndpoint)
-    return (
-        isunbounded(a) ||
-        isunbounded(b) ||
-        a.endpoint < b.endpoint
-    )
+function Base.isequal(a::AbstractEndpoint, b::AbstractEndpoint)
+    if isunbounded(a) || isunbounded(b)
+        return (isunbounded(a) && isunbounded(b) && isleft(a) == isleft(b))
+    elseif !isequal(a.endpoint, b.endpoint)
+        return false
+    elseif isleft(a) == isleft(b)
+        return isclosed(a) == isclosed(b) 
+    else 
+        return isclosed(a) && isclosed(b)
+    end
 end
 
-function Base.isless(a::RightEndpoint, b::LeftEndpoint)
-    return (
-        !isunbounded(a) && !isunbounded(b) &&
-        (
-            a.endpoint < b.endpoint ||
-            a.endpoint == b.endpoint && !(isclosed(a) && isclosed(b))
-        )
-    )
+function Base.isless(a::AbstractEndpoint, b::AbstractEndpoint)
+    if isleft(a) == isleft(b)
+        if isunbounded(a) || isunbounded(b)
+            return isleft(a) ? 
+                isbounded(a) < isbounded(b) :
+                isunbounded(a) < isunbounded(b)
+        elseif a.endpoint == b.endpoint 
+            return offset_value(a) < offset_value(b)
+        else
+            a.endpoint < b.endpoint
+        end
+    else
+        if isunbounded(a)
+            return isleft(a)
+        elseif isunbounded(b)
+            return !isleft(b)
+        elseif a.endpoint == b.endpoint
+            return !(isclosed(a) && isclosed(b)) && isleft(a) < isleft(b)
+        else
+            a.endpoint < b.endpoint
+        end
+    end
 end
 
 # Comparisons between Scalars and Endpoints
-Base.:(==)(a, b::Endpoint) = a == b.endpoint && isclosed(b)
-Base.:(==)(a::Endpoint, b) = b == a
-
-function Base.isless(a, b::LeftEndpoint)
-    return (
-        !isunbounded(b) && (
-            a < b.endpoint ||
-            a == b.endpoint && !isclosed(b)
+Base.:(==)(a::Number, b::Endpoint) = a == b.endpoint && isclosed(b)
+Base.:(==)(a::Endpoint, b::Number) = b == a
```

Comment:
Endpoints are not restricted to numeric types. We may need to compare against `Date`s, `DateTime`s, `Char`s etc.